linkerd/proxy/transport/src/orig_dst.rs

```diff
@@ -0,0 +1,191 @@
+use crate::{
+    addrs::*,
+    listen::{self, Bind, Bound},
+};
+use futures::prelude::*;
+use linkerd_io as io;
+use linkerd_stack::Param;
+use std::pin::Pin;
+use tokio::net::TcpStream;
+
+#[derive(Copy, Clone, Debug, Default)]
+pub struct BindWithOrigDst<B = listen::BindTcp> {
+    inner: B,
+}
+
+#[derive(Clone, Debug)]
+pub struct Addrs<A = listen::Addrs> {
+    pub inner: A,
+    pub orig_dst: OrigDstAddr,
+}
+
+// === impl Addrs ===
+
+impl<A> Param<OrigDstAddr> for Addrs<A> {
+    fn param(&self) -> OrigDstAddr {
+        self.orig_dst
+    }
+}
+
+impl<A> Param<Remote<ClientAddr>> for Addrs<A>
+where
+    A: Param<Remote<ClientAddr>>,
+{
+    fn param(&self) -> Remote<ClientAddr> {
+        self.inner.param()
+    }
+}
+
+impl<A> Param<Local<ServerAddr>> for Addrs<A>
+where
+    A: Param<Local<ServerAddr>>,
+{
+    fn param(&self) -> Local<ServerAddr> {
+        self.inner.param()
+    }
+}
+
+// === impl WithOrigDst ===
+
+impl<B> From<B> for BindWithOrigDst<B> {
+    fn from(inner: B) -> Self {
+        Self { inner }
+    }
+}
+
+impl<T, B> Bind<T> for BindWithOrigDst<B>
+where
+    B: Bind<T, Io = TcpStream> + 'static,
+{
+    type Addrs = Addrs<B::Addrs>;
+    type Io = TcpStream;
+    type Incoming =
+        Pin<Box<dyn Stream<Item = io::Result<(Self::Addrs, TcpStream)>> + Send + Sync + 'static>>;
+
+    fn bind(self, t: &T) -> io::Result<Bound<Self::Incoming>> {
+        let (addr, incoming) = self.inner.bind(t)?;
+
+        let incoming = incoming.map(|res| {
+            let (inner, tcp) = res?;
+            let orig_dst = orig_dst_addr(&tcp)?;
+            let addrs = Addrs { inner, orig_dst };
+            Ok((addrs, tcp))
+        });
+
+        Ok((addr, Box::pin(incoming)))
```

Comment:
btw, I tried to get rid of the box, but this closure captures `B` (types -- not values) so we get a _expected fn pointer found closure_ compiler error. if you know of a way to avoid the box, by all means...